contracts/SwapFlashLoan.sol

```diff
@@ -0,0 +1,169 @@
+        uint8[] memory decimals,
+        string memory lpTokenName,
+        string memory lpTokenSymbol,
+        uint256 _a,
+        uint256 _fee,
+        uint256 _adminFee,
+        uint256 _withdrawFee,
+        IAllowlist _allowlist
+    )
+        public
+        Swap(
+            _pooledTokens,
+            decimals,
+            lpTokenName,
+            lpTokenSymbol,
+            _a,
+            _fee,
+            _adminFee,
+            _withdrawFee,
+            _allowlist
+        )
+    {
+        flashLoanFeeBips = 100; // 100bps
+        flashLoanProtocolFeeBips = 5000; // 5000bps
+    }
+
+    /*** STATE MODIFYING FUNCTIONS ***/
+
+    /**
+     * @notice Borrow the specified token from this pool for this transaction only. This function will call
+     * `IFlashLoanReceiver(receiver).executeOperation` and the `receiver` must return the full amount of the token
+     * and the associated fee by the end of the callback transaction. If the conditions are not met, this call
+     * is reverted.
+     * @param receiver the address of the receiver of the token. This address must implement the IFlashLoanReceiver
+     * interface and the callback function `executeOperation`.
+     * @param token the protocol fee in bps to be applied on the total flash loan fee
+     * @param amount the total amount to borrow in this transaction
+     * @param params optional data to pass along to the callback function
+     */
+    function flashLoan(
+        address receiver,
+        IERC20 token,
+        uint256 amount,
+        bytes memory params
+    ) external nonReentrant {
+        uint8 tokenIndex = getTokenIndex(address(token));
+        uint256 availableLiquidityBefore = token.balanceOf(address(this));
+        uint256 protocolBalanceBefore =
+            availableLiquidityBefore.sub(swapStorage.balances[tokenIndex]);
+        require(
+            amount > 0 && availableLiquidityBefore >= amount,
+            "invalid amount"
+        );
+
+        // Calculate the additional amount of tokens the pool should end up with
+        uint256 amountFee = amount.mul(flashLoanFeeBips).div(10000);
+        // Calculate the portion of the fee that will go to the protocol
+        uint256 protocolFee =
+            amountFee.mul(flashLoanProtocolFeeBips).div(10000);
+        require(
+            amountFee > 0,
+            "The requested amount is too small for a flashLoan"
+        );
+
+        // Transfer the requested amount of tokens
+        token.safeTransfer(receiver, amount);
+
+        // Execute callback function on receiver
+        IFlashLoanReceiver(receiver).executeOperation(
+            address(this),
+            address(token),
+            amount,
+            amountFee,
+            params
+        );
+
+        uint256 availableLiquidityAfter =
+            token.balanceOf(address(this)).sub(protocolBalanceBefore);
+        require(
+            availableLiquidityAfter == availableLiquidityBefore.add(amountFee),
```

Comment:
could this be `>=` ? would that correctly allow the flash loan recipient to overpay and give the excess to the pool participants?

Reply:
I didn't want to be too greedy but that works for me 😛  Thoughts on this @alphastorm ?

Reply:
sounds good to me, there might be some generous arbitrageurs out there 😜